feat: `asSpecifiedCallsOnlyV14` option

TarikGul opened the pull request:
closes: https://github.com/paritytech/txwrapper-core/issues/111

Feature:

Currently, we have `asCallsOnly` for reducing the size of the metadata to only calls, the types, and extrinsic information. This PR adds an option called `asSpecifiedCallsOnlyV14` for `createMetadata` which is represented as an array of strings. Each string representing a pallet. All calls within that pallet will be stored within this generated metadata, and it will strip all the lookup type information that isnt necessary for the required calls. 

Reduces the metadata size by about ~~92-97%~~ 70% and above depending on the inputted pallets to be kept within the metadata.

TarikGul committed:
add resolveJsonModule for mock testing

TarikGul committed:
add mock response for toSpecifiedCallsOnlyV14

TarikGul committed:
toSpecifiedCallsOnlyV14

TarikGul committed:
add option to createMetadata

TarikGul committed:
fix options

TarikGul committed:
remove unnecessaary inline docs

TarikGul committed:
update comments

TarikGul committed:
fix tsc build

TarikGul committed:
lint

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+
+	// Variant types
+	if (def.isVariant) {
+		// Can this be more efficient
```
I dont think its worth over thinking this.

TarikGul replied:
BUT, at the same time would be nice to not have a nested loop here.

jsdw replied:
I don't think you can avoid this; every variant has fields and every field will have its own type, and you care about them all

TarikGul replied:
Yea exactly, you're 100% correct.

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+	name: Text;
+}
+
+/**
```
Inline docs?

TarikGul committed:
add asSpecifiedCallsOnlyV14 to txwrapper-polkadot

TarikGul committed:
add asSpecifiedCallsOnlyV14 to txwrapper-regsitry

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+	// Primitive types
+	// NOTE: This check might not be necessary
+	if (def.isPrimitive) {
+		// for primitive types, we only need the root id. If this isnt necessary
```
Follow up

TarikGul replied:
Oh my bad, I worded that wrong :) I meant if we don't need to check for the primitive type, then we can remove the `id` from this specific functions input params.

jsdw replied:
I guess you still need to either check for the primitive types here and add them to the set, or have special logic below which always keeps any primitive type in the registry when it's working out what to filter?

TarikGul replied:
Ahhh so actually we dont need to check for primitives at all. (Things are a little clearer after re-reading a few times). Reason being, when we traverse the types we do a check:

```typescript
if (typeHashMap.has(id.toString())) {
    // Traverse the typedef for its required sub types
    findLookupIdsInTypeDef(type, id, typeHashMap);
}
```

So if it's a primitive, it's id is already in the set.

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+	});
+
+	// Add all necessary type id's to the set from each type definition
+	// It reverses the array so correctly traverse the types, making sure primitive
```
```suggestion
	// It reverses the array to correctly traverse the types, making sure primitive
```

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+	// It reverses the array so correctly traverse the types, making sure primitive
+	// types are last.
+	latestMetadata.lookup.types.reverse().forEach(({ id, type }) => {
+		if (typeHashMap.has(id.toString())) {
```
Are there any optimizations we can make here?

jsdw replied:
A small one would be to just iterate the `latestMetadata.lookup.types` list in reverse (`for (let i = types.length - 1; i >= 0; i--)` type thing).

I guess the assumption is that later types only ever reference earlier types (which makes sense, but I don't know if it's a guarantee?). In that case this seems like quite an effective way to go about it to me!

TarikGul replied:
Thank you for putting that in words for me, I couldnt quite verbalize it when I was going over this last night. When i wrote that the assumption was the types, would only reference their sub-types linearly. This is definitely a check that needs to be confirmed.

TarikGul replied:
It's not guaranteed. So I need ti rewrite his with some type of queue

TarikGul commented on packages/txwrapper-core/src/core/metadata/createMetadata.spec.ts:
```diff
@@ -52,4 +58,20 @@ describe('createMetadata', () => {
 			metadataAsCallsBuffer.byteLength
 		);
 	});
+
```
Yea I was thinking about this last night as well, some reliable test to confirm that all necessary types are added would be really nice.

jsdw replied:
Yeah I agree; if there was a stupid "dumb" version then it would be reasonable, but if it's just reimplementing the same again it doesn't seem worth while.

Would it be feasible to have a couple of tests that start with a hand crafted metadata representation that contains just a few types and couple of pallets, and make sure that the expected types are removed if we want to only keep one of those pallets?

TarikGul replied:
Yea totally, I could build that out for sure! What I'll do is take the current json response we expect, and add some types to it that dont fit it. Then we should expect it to strip all the unnecessary types.

TarikGul replied:
So actually have been working through creating more tests, I think the current one is pretty sufficient. The type generation required to make a small size set of mock metadata is a bit, and requires a little overhead, which I think the current test already handles. Lmk what you think. (I manually worked through the metadata tree response that we have in the current tests and it all checks out)

TarikGul committed:
Update packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,234 @@
+	if (def.isComposite) {
+		def.asComposite.fields.forEach((f) => {
+			const lookupId = f.type.toString();
+			if (!set.has(lookupId)) set.add(lookupId);
```
Yea I was thinking about this as well, all values in a set will be unique regardless, and the has is just another read..

TarikGul replied:
fixed

TarikGul committed:
remove unnecessary set.has checks

TarikGul committed:
remove the check for primitives

TarikGul committed:
fix the algorithm and add a stack - while loop

TarikGul committed:
fix small inline grumbles

TarikGul committed:
lint

jsdw commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,289 @@
+	if (def.isComposite) {
+		def.asComposite.fields.forEach((f) => {
+			const lookupId = f.type.toString();
+			if (!cache.has(lookupId)) {
+				cache.add(lookupId);
+				stack.push(lookupId);
+			}
```
Just a tiny thing; perhaps you could add a little function inside findLookupIdsInTypeDef like:

```
function tryAddType(lookupId: Type) { // whatever `Type` should be..)
    const lookupStr = lookupId.toString();
    if (!cache.has(lookupId) {
        cache.add(lookupId);
        stack.push(lookupId);
    }
}
```

and then reuse it everywhere instead of all of the string convert+cache-add+stack-add bits :)

TarikGul replied:
Rad fixed this portion!

jsdw commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,289 @@
+	 * if they dont exist. It will continue to increase the stack if there are types
+	 * yet to be found that are necessary.
+	 */
+	while (stack.length > 0) {
```
I guess until this line you don't need a stack _and_ a Set; the above logic (and `findLookupIdsInPallet`) could just push types into a set, and then just before here you could turn that set into a stack, and run this look until the stack runs out. (just because then you only have to add types to a set and not keep a stack in sync for all of the above :))

jsdw commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,289 @@
+	 */
+	while (stack.length > 0) {
+		// Typecast is fine here since the array will never be empty if we are here
+		const item = stack.pop() as string;
```
I think you can also do `stack.pop()!` to assert the same thing?

jsdw replied:
Aah, if you want to keep the link rule then perhaps how you had it was best!

jsdw reviewed (commented):
Nice, I like the stack based approach!

TarikGul committed:
remove typecast, and assert stack.pop() with !

TarikGul commented on packages/txwrapper-core/src/core/metadata/toSpecifiedCallsOnlyV14.ts:
```diff
@@ -0,0 +1,289 @@
+	cache: Set<unknown>,
+	stack: string[]
+): void => {
+	// Storage types
```
So actually for this, since the stack isnt going to be established until right before the loop, we can remove all the 

```
if (!cache.has(lookupId)) {
    cache.add(lookupId);
    stack.push(lookupId);
}
```

and replace it with just `cache.add(lookupId)`

TarikGul committed:
copyright credit for polkadot-js

TarikGul committed:
refactor stack generation, and add tryAddType to clean code

TarikGul committed:
inline docs for tryAddType

jsdw reviewed (approved):
Looks great :)

TarikGul committed:
add toSpecifiedCallsOnlyV14 to metadata exports

TarikGul committed:
revert stack.pop to typecast

TarikGul committed:
fix source ref

giorgiapellizzari commented:
Hello! When can we hope for the PR to be approved? Thanks!

TarikGul committed:
add params lookup

TarikGul committed:
change mock test to use balances pallet

TarikGul committed:
change test to test against balance pallet

TarikGul committed:
strip all docs

TarikGul committed:
fix grumbles

TarikGul committed:
Merge branch 'main' of github.com:paritytech/txwrapper-core into tarik-metadata-min

TarikGul committed:
add example of asSpecifiedCallsOnlyV14 options

TarikGul commented:
For clarity sake, I missed adding the `params` sub types which added a little bit to the metadata. Once I added those in, I also added to the logic to remove all docs from every type. 

I added an example to `txwrapper-examples/options` that shows how to use `asSpecifiedToCallsOnlyV14`. The given output for stripping the metadata to only the `balances` pallet is as such. 

```
Original metadata byte size: 664052
Specified Calls Only V14 metadata byte size: 126430
Metadata size decreased by 80.96%
```

It will obviously vary per pallet. I also made sure to test the following with live transactions, and everything worked as should.

TarikGul commented:
The last thing that needs to be worked out, is `decode` and all instances where `createMetadata` is called accepts the `asSpecifiedCallsOnlyV14` option.

TarikGul committed:
add asSpecifiedCallsOnlyV14 to OptionsWithMeta

TarikGul committed:
cleanup example

TarikGul committed:
add extra check for toSpecifiedCallsOnlyV14

TarikGul committed:
adjust the tests to include system
